src/prolong_restrict/pr_loops.hpp

```diff
@@ -92,20 +126,76 @@ ProlongationRestrictionLoop(const BufferCache_t &info, const Idx_t &buffer_idxs,
-          int sk, ek, sj, ej, si, ei;
-          GetLoopBoundsFromBndInfo<DIM>(info(buf), ckb.s, cjb.s, sk, ek, sj, ej, si, ei);
-          par_for_inner(inner_loop_pattern_ttr_tag, team_member, 0, info(buf).Nt - 1, 0,
-                        info(buf).Nu - 1, 0, info(buf).Nv - 1, sk, ek, sj, ej, si, ei,
-                        [&](const int t, const int u, const int v, const int k,
-                            const int j, const int i) {
-                          Stencil::template Do<DIM>(
-                              t, u, v, k, j, i, ckb, cjb, cib, kb, jb, ib,
-                              info(buf).coords, info(buf).coarse_coords,
-                              &(info(buf).coarse), &(info(buf).fine));
-                        });
+          using TE = TopologicalElement;
+          if (info(buf).fine.topological_type == TopologicalType::Cell)
+            IterateInnerProlongationRestrictionLoop<DIM, Stencil, TE::C>(
+                team_member, buf, info, ckb, cjb, cib, kb, jb, ib);
+          if (info(buf).fine.topological_type == TopologicalType::Face)
+            IterateInnerProlongationRestrictionLoop<DIM, Stencil, TE::FX, TE::FY, TE::FZ>(
+                team_member, buf, info, ckb, cjb, cib, kb, jb, ib);
+          if (info(buf).fine.topological_type == TopologicalType::Edge)
+            IterateInnerProlongationRestrictionLoop<DIM, Stencil, TE::EXY, TE::EXZ,
+                                                    TE::EYZ>(team_member, buf, info, ckb,
+                                                             cjb, cib, kb, jb, ib);
+          if (info(buf).fine.topological_type == TopologicalType::Node)
+            IterateInnerProlongationRestrictionLoop<DIM, Stencil, TE::NXYZ>(
+                team_member, buf, info, ckb, cjb, cib, kb, jb, ib);
         }
       });
 }
+
+template <int DIM, class Stencil, TopologicalElement FEL, TopologicalElement CEL>
+inline void
+InnerHostProlongationRestrictionLoop(std::size_t buf, const BufferCacheHost_t &info,
+                                     const IndexRange &ckb, const IndexRange &cjb,
+                                     const IndexRange &cib, const IndexRange &kb,
+                                     const IndexRange &jb, const IndexRange &ib) {
+  const auto &idxer = info(buf).prores_idxer[static_cast<int>(CEL)];
+  auto coords = info(buf).coords;
+  auto coarse_coords = info(buf).coarse_coords;
+  auto coarse = info(buf).coarse;
+  auto fine = info(buf).fine;
+  par_for(
+      DEFAULT_LOOP_PATTERN, "ProlongateOrRestrictCellCenteredValues", DevExecSpace(), 0,
+      0, 0, 0, 0, idxer.size() - 1, KOKKOS_LAMBDA(const int, const int, const int ii) {
+        const auto [t, u, v, k, j, i] = idxer(ii);
+        Stencil::template Do<DIM, FEL, CEL>(t, u, v, k, j, i, ckb, cjb, cib, kb, jb, ib,
+                                            coords, coarse_coords, &coarse, &fine);
+      });
+}
+
+template <int DIM, class Stencil, TopologicalElement... ELs, class... Args>
```

Comment:
Why is `TopologicalElement` variatic here? It doesn't seem like you ever use more than one in the calls below? Or am I missing something in how this parameter pack is working?

Reply:
This was just a way of making a compile time loop over topological elements (e.g. the three faces if you are face field). Probably not necessary and confusing, so better to remove it.